File: MB-Secretaria/observability/trace-engine.ts
import { InMemoryObservabilityEventBus } from "./event-bus";
import { buildRequestLineage } from "./request-lineage";
import { buildStructuredLog } from "./structured-logger";
import { createTraceContext } from "./trace-context";
import type { ObservabilityEvent, RequestLineage, TraceContext } from "./types";

export class TraceEngine {
  constructor(private readonly eventBus = new InMemoryObservabilityEventBus()) {}

  startTrace(input: Parameters<typeof createTraceContext>[0]): TraceContext {
    return createTraceContext(input);
  }

  record(input: Parameters<typeof buildStructuredLog>[0]): ObservabilityEvent {
    const event = buildStructuredLog(input);
    this.eventBus.publish(event);
    return event;
  }

  events(): ObservabilityEvent[] {
    return this.eventBus.list();
  }

  lineage(trace_id: string): RequestLineage {
    return buildRequestLineage(trace_id, this.events());
  }
}

File: MB-Whatsapp/observability_py/trace_engine.py
from typing import Any

from .event_bus import InMemoryObservabilityEventBus
from .request_lineage import build_request_lineage
from .structured_logger import build_structured_log
from .trace_context import create_trace_context
from .types import ObservabilityEvent, ObservabilityLayer, ObservabilitySeverity, RequestLineage, TraceContext


class TraceEngine:
    def __init__(self, event_bus: InMemoryObservabilityEventBus | None = None) -> None:
        self.event_bus = event_bus or InMemoryObservabilityEventBus()

    def start_trace(self, tenant_id: str, request_type: str, source_layer: str, parent_trace_id: str | None = None) -> TraceContext:
        return create_trace_context(tenant_id, request_type, source_layer, parent_trace_id)

    def record(
        self,
        trace: TraceContext,
        layer: ObservabilityLayer,
        event_type: str,
        severity: ObservabilitySeverity = "info",
        payload_summary: dict[str, Any] | None = None,
        safety_flags: list[str] | None = None,
    ) -> ObservabilityEvent:
        event = build_structured_log(trace, layer, event_type, severity, payload_summary, safety_flags)
        self.event_bus.publish(event)
        return event

    def events(self) -> list[ObservabilityEvent]:
        return self.event_bus.list()

    def lineage(self, trace_id: str) -> RequestLineage:
        return build_request_lineage(trace_id, self.events())
